File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/08_greedy_algorithms/kotlin/01_set_covering.kt
var statesNeeded = setOf("mt", "wa", "or", "id", "nv", "ut", "ca", "az")
val stations = mapOf(
    "kone" to setOf("id", "nv", "ut"),
    "ktwo" to setOf("wa", "id", "mt"),
    "kthree" to setOf("or", "nv", "ca"),
    "kfour" to setOf("nv", "ut"),
    "kfive" to setOf("ca", "az")
)

fun main() {
    val finalStations = mutableSetOf<String>()
    while (statesNeeded.isNotEmpty()) {
        var bestStation: String? = null
        var statesCovered = setOf<String>()
        stations.forEach { (station, states) ->
            val covered = statesNeeded.intersect(states)
            if (covered.size > statesCovered.size) {
                bestStation = station
                statesCovered = covered
            }
        }
        statesNeeded = statesNeeded - statesCovered
        bestStation?.let { finalStations.add(it) }
    }
    println(finalStations)
}

File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/08_greedy_algorithms/php/01_set_covering.php
<?php
// You need to install Data Structures (DS) extension. 
// More details here - http://php.net/manual/en/book.ds.php
use \Ds\Set;

$statesNeeded = new Set(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"]);
$stations = [];
$stations["kone"] = new Set(["id", "nv", "ut"]);
$stations["ktwo"] = new Set(["wa", "id", "mt"]);
$stations["kthree"] = new Set(["or", "nv", "ca"]);
$stations["kfour"] = new Set(["nv", "ut"]);
$stations["kfive"] = new Set(["ca", "az"]);
$finalStations = new Set();

while (!$statesNeeded->isEmpty()) {
    $bestStation = null;
    $statesCovered = new Set();
    foreach (array_keys($stations) as $station) {
        $states = $stations[$station];
        $covered = new Set($statesNeeded);
        $covered = $covered->filter(function ($value) use ($states) {
            return $states->contains($value);
        });
        if ($covered->count() > $statesCovered->count()) {
            $bestStation = $station;
            $statesCovered = $covered;
        }
    }
    $statesNeeded = new Set($statesNeeded);
    $statesNeeded = $statesNeeded->filter(function ($value) use ($statesCovered) {
        return !$statesCovered->contains($value);
    });
    $finalStations->add($bestStation);
}

print_r($finalStations); // ["kone", "ktwo", "kthree", "kfive"]